File: src/main/java/com/somefrills/misc/SkyblockData.java
package com.somefrills.misc;

import com.somefrills.events.*;
import meteordevelopment.orbit.EventHandler;
import meteordevelopment.orbit.EventPriority;
import net.minecraft.client.network.PlayerListEntry;
import net.minecraft.network.packet.s2c.query.PingResultS2CPacket;
import net.minecraft.scoreboard.*;
import net.minecraft.util.Formatting;
import net.minecraft.util.Util;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Pattern;

import static com.somefrills.Main.eventBus;
import static com.somefrills.Main.mc;

public class SkyblockData {
    private static final Pattern scoreRegex = Pattern.compile("Team Score: [0-9]* (.*)");
    private static String location = "";
    private static Area area = Area.UNKNOWN;
    private static boolean inSkyblock = false;
    private static boolean instanceOver = false;
    private static List<String> tabListLines = new ArrayList<>();
    private static List<String> lines = new ArrayList<>();
    private static boolean showPing = false;
    private static boolean tabListDirty = true;
    private static boolean scoreboardDirty = true;

    /**
     * Returns the current location from the scoreboard, such as "⏣ Your Island". The location prefix is not omitted.
     */
    public static String getLocation() {
        return location;
    }

    /**
     * Returns the current area from the tab list, such as "Area: Private Island". The area/dungeon prefix is omitted.
     */
    public static Area getArea() {
        return area;
    }

    public static boolean isInSkyblock() {
        return inSkyblock;
    }

    public static boolean isInInstance() {
        return Utils.isInDungeons() || Utils.isInKuudra();
    }

    public static boolean isInstanceOver() {
        return instanceOver;
    }

    public static List<String> getTabListLines() {
        return tabListLines;
    }

    public static List<String> getLines() {
        return lines;
    }

    public static void showPing() {
        showPing = true;
        Utils.sendPingPacket();
    }

    public static void markTabListDirty() {
        tabListDirty = true;
    }

    private static void updateTabListIfDirty() {
        List<String> lines = new ArrayList<>();
        if (mc.inGameHud == null || mc.player == null || mc.player.networkHandler == null) return;
        var playerListHud = mc.inGameHud.getPlayerListHud();
        if (playerListHud == null) return;

        for (PlayerListEntry entry : playerListHud.collectPlayerEntries()) {
            if (entry == null || entry.getDisplayName() == null) continue;

            String name = Utils.toPlain(entry.getDisplayName()).trim();
            if (name.isEmpty()) continue;

            if (name.startsWith("Area: ") || name.startsWith("Dungeon: ")) {
                String areaStr = name.split(":", 2)[1].trim();
                Area newArea = Area.fromString(areaStr);
                if (newArea != area) {
                    area = newArea;
                    eventBus.post(new AreaChangeEvent(area));
                }
            }
            lines.add(name);
        }
        tabListLines = lines;
    }

    private static void updateTabList() {
        if (tabListDirty) {
            updateTabListIfDirty();
            tabListDirty = false;
            eventBus.post(new TabListUpdateEvent(tabListLines));
        }
    }

    public static void updateObjective() {
        if (mc.player != null) {
            Scoreboard scoreboard = mc.player.networkHandler.getScoreboard();
            ScoreboardObjective objective = scoreboard.getObjectiveForSlot(ScoreboardDisplaySlot.FROM_ID.apply(1));
            if (objective != null) {
                inSkyblock = Utils.toPlain(objective.getDisplayName()).contains("SKYBLOCK");
            }
        }
    }

    public static void markScoreboardDirty() {
        scoreboardDirty = true;
    }

    private static void updateScoreboardIfDirty() {
        if (mc.player != null) {
            List<String> currentLines = new ArrayList<>();
            Scoreboard scoreboard = mc.player.networkHandler.getScoreboard();
            ScoreboardObjective objective = scoreboard.getObjectiveForSlot(ScoreboardDisplaySlot.FROM_ID.apply(1));
            for (ScoreHolder scoreHolder : scoreboard.getKnownScoreHolders()) {
                if (scoreboard.getScoreHolderObjectives(scoreHolder).containsKey(objective)) {
                    Team team = scoreboard.getScoreHolderTeam(scoreHolder.getNameForScoreboard());
                    if (team != null) {
                        String line = Formatting.strip(team.getPrefix().getString() + team.getSuffix().getString()).trim();
                        if (!line.isEmpty()) {
                            if (line.startsWith(Utils.Symbols.zone) || line.startsWith(Utils.Symbols.zoneRift)) {
                                location = line;
                            }
                            if (Utils.isInKuudra() && !instanceOver) {
                                instanceOver = line.startsWith("Instance Shutdown");
                            }
                            currentLines.add(line);
                        }
                    }
                }
            }
            lines = currentLines;
        }
    }

    private static void updateScoreboard() {
        if (scoreboardDirty) {
            updateScoreboardIfDirty();
            scoreboardDirty = false;
        }
    }

    @EventHandler
    private static void onChat(ChatMsgEvent event) {
        if (Utils.isInDungeons()) {
            if (!instanceOver && scoreRegex.matcher(event.messagePlain.trim()).matches()) {
                instanceOver = true;
            }
        }
    }

    @EventHandler
    private static void onJoinServer(ServerJoinEvent event) {
        instanceOver = false;
        inSkyblock = false;
        location = "";
        area = Area.UNKNOWN;
        lines.clear();
    }

    @EventHandler
    private static void onPing(ReceivePacketEvent event) {
        if (showPing && event.packet instanceof PingResultS2CPacket(long startTime)) {
            Utils.infoFormat("§aPing: §f{}ms", Util.getMeasuringTimeMs() - startTime);
            showPing = false;
        }
    }

    @EventHandler(priority = EventPriority.HIGHEST)
    private static void onWorldTick(TickEventPost event) {
        updateTabList();
        updateScoreboard();
    }
}

File: src/main/kotlin/com/somefrills/misc/SkyblockData.kt
package com.somefrills.misc

import com.somefrills.Main
import com.somefrills.Main.mc
import com.somefrills.events.*
import com.somefrills.utils.*
import meteordevelopment.orbit.EventHandler
import meteordevelopment.orbit.EventPriority
import net.minecraft.ChatFormatting
import net.minecraft.network.protocol.ping.ClientboundPongResponsePacket
import net.minecraft.world.scores.DisplaySlot
import java.util.regex.Pattern


object SkyblockData {
    private val scoreRegex: Pattern = Pattern.compile("Team Score: [0-9]* (.*)")

    /**
     * Returns the current location from the scoreboard, such as "⏣ Your Island". The location prefix is not omitted.
     */
    @JvmStatic
    var location: String = ""
        private set

    /**
     * Returns the current area from the tab list, such as "Area: Private Island". The area/dungeon prefix is omitted.
     */
    @JvmStatic
    var area: Area = Area.UNKNOWN
        private set

    @JvmStatic
    var isInSkyblock: Boolean = false
        private set

    @JvmStatic
    var isInstanceOver: Boolean = false
        private set

    @JvmStatic
    var tabListLines: MutableList<String> = ArrayList()
        private set

    @JvmStatic
    var lines: MutableList<String> = ArrayList()
        private set
    private var showPing = false
    private var tabListDirty = true
    private var scoreboardDirty = true

    fun showPing() {
        showPing = true
        NetworkUtils.sendPingPacket()
    }

    @JvmStatic
    fun markTabListDirty() {
        tabListDirty = true
    }

    private fun updateTabListIfDirty() {
        val lines: MutableList<String> = ArrayList()
        if (mc.gui == null || mc.player == null ) return
        for (entry in  mc.gui.tabList.playerInfos) {
            val displayName = entry.tabListDisplayName?.toPlain() ?: continue
            val name = displayName.trim { it <= ' ' }
            if (name.isEmpty()) continue

            if (name.startsWith("Area: ") || name.startsWith("Dungeon: ")) {
                val areaStr = name.split(":".toRegex(), limit = 2).toTypedArray()[1].trim { it <= ' ' }
                val newArea = Area.fromString(areaStr)
                if (newArea != area) {
                    area = newArea
                    Main.LOGGER.info("Detected area change: {}", area.displayName)
                    AreaChangeEvent(area).post()
                }
            }
            lines.add(name)
        }
        tabListLines = lines
    }

    private fun updateTabList() {
        if (tabListDirty) {
            updateTabListIfDirty()
            tabListDirty = false
            TabListUpdateEvent(tabListLines).post()
        }
    }

    @JvmStatic
    fun updateObjective() {
        if (mc.player == null) return
        val scoreboard = mc.player?.connection?.scoreboard() ?: return
        val objective = scoreboard.getDisplayObjective(DisplaySlot.BY_ID.apply(1))
        if (objective != null) {
            isInSkyblock = objective.displayName.toPlain().contains("SKYBLOCK")
        }
    }

    @JvmStatic
    fun markScoreboardDirty() {
        scoreboardDirty = true
    }

    private fun updateScoreboardIfDirty() {
        if (mc.player != null) {
            val currentLines = ArrayList<String>()
            val scoreboard = mc.player?.connection?.scoreboard() ?: return
            val objective = scoreboard.getDisplayObjective(DisplaySlot.BY_ID.apply(1))
            for (scoreHolder in scoreboard.trackedPlayers) {
                if (!scoreboard.listPlayerScores(scoreHolder).containsKey(objective)) continue
                val team = scoreboard.getPlayersTeam(scoreHolder.scoreboardName)
                if (team != null) {
                    val line = ChatFormatting.stripFormatting(
                        team.playerPrefix.string + team.playerSuffix.string
                    )!!.trim { it <= ' ' }
                    if (!line.isEmpty()) {
                        if (line.startsWith(Symbols.zone) || line.startsWith(Symbols.zoneRift)) {
                            location = line
                        }
                        if (SkyblockUtils.isInKuudra() && !isInstanceOver) {
                            isInstanceOver = line.startsWith("Instance Shutdown")
                        }
                        currentLines.add(line)
                    }
                }
            }
            lines = currentLines
        }
    }

    private fun updateScoreboard() {
        if (scoreboardDirty) {
            updateScoreboardIfDirty()
            scoreboardDirty = false
        }
    }

    @EventHandler
    @JvmStatic
    private fun onChat(event: ChatMsgEvent) {
        if (!SkyblockUtils.isInDungeons()) return
        if (!isInstanceOver && scoreRegex.matcher(event.plainMessage.trim { it <= ' ' }).matches()) {
            isInstanceOver = true
        }
    }

    @EventHandler
    @JvmStatic
    private fun onJoinServer(event: ServerJoinEvent) {
        isInstanceOver = false
        isInSkyblock = false
        location = ""
        area = Area.UNKNOWN
        lines.clear()
    }

    @EventHandler
    @JvmStatic
    private fun onPing(event: ReceivePacketEvent) {
        if (!showPing) return
        event.packet.let {
            if (it is ClientboundPongResponsePacket) {
                ChatUtils.infoFormat("§aPing: §f{}ms", it.time)
                showPing = false
            }
        }
    }

    @EventHandler(priority = EventPriority.HIGHEST)
    @JvmStatic
    private fun onWorldTick(event: TickEventPost) {
        updateTabList()
        updateScoreboard()
    }
}